python/dgl/backend/pytorch/sparse.py

```diff
@@ -145,6 +145,84 @@ def backward(ctx, dZ):
         return None, None, None, dX, dY
 
 
+class GSpMM_hetero(th.autograd.Function):
+    @staticmethod
+    @custom_fwd(cast_inputs=th.float16)
+    def forward(ctx, g, op, reduce_op, *X): # X = X + Y
+        out, (argX, argY) = _gspmm_hetero(g, op, reduce_op, X)
+        ctx.backward_cache = g, op, reduce_op
+        ctx.save_for_backward(*X, argX, argY)
+        return out
+
+    @staticmethod
+    @custom_bwd
+    def backward(ctx, *dZ):
+        g, op, reduce_op = ctx.backward_cache 
+        X = ctx.saved_tensors[:-2]
+        argX=ctx.saved_tensors[-2]  
+        argY=ctx.saved_tensors[-1]
+        num_ntypes = g._graph.number_of_ntypes()
+        X ,Y = X[:num_ntypes], X[num_ntypes:]
+
+
+        if op != 'copy_rhs' and any([x is not None for x in X]):
```

Comment:
I see only a certain branch (i.e., when `op == 'copy_lhs'`). I suppose you will update the rest in future PRs. If that's the case, please have an explicit check before. Something like:
```python
if g._graph.number_of_etypes() > 1 and (op, reduce_op) not in [('copy_lhs', 'sum'), ('copy_rhs', 'sum')]:
    raise NotImplementedError('... some error msg ...')

# begin the main logic ...
if op != 'copy_rhs' and any([x is not None for x in X]):
```

Reply:
We have this check at the beginning of message passing function in python/dgl/heterograph.py. We can consider deleting the part since the code will never reach this point.